### PR TITLE
TEALv3 - implement bit opcodes

### PR DESCRIPTION
**Pull request by amityadav0**
- Add `SetBit` and `GetBit` opcodes
- Add tests

### CODE REVIEW

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 1]
```diff
@@ -2259,3 +2259,97 @@ export class Swap extends Op {
...
   }
 }
+
+/**
+ * bit indexing begins with low-order bits in integers.
+ * Setting bit 4 to 1 on the integer 0 yields 16 (int 0x0010, or 2^4).
+ * Indexing begins in the first bytes of a byte-string
+ * (as seen in getbyte and substring). Setting bits 0 through 11 to 1
+ * in a 4 byte-array of 0s yields byte 0xfff00000
+ */
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:2269]: please also add in description what this opcodes pops & pushes to stack

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 2]
```diff
@@ -2259,3 +2259,97 @@ export class Swap extends Op {
...
+    if (bit > 1n) {
+      throw new RuntimeError(RUNTIME_ERRORS.TEAL.SET_BIT_VALUE_ERROR, { line: this.line });
+    }
+
+    if (typeof target === "bigint") {
+      this.assert64BitIndex(index, this.line);
+      const binaryStr = target.toString(2);
+      const binaryArr = [...(binaryStr.padStart(64, "0"))];
+      const size = binaryArr.length;
+      binaryArr[size - Number(index) - 1] = bit === 0n ? "0" : "1";
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:2298]:
```suggestion
      binaryArr[size - Number(index) - 1] = (bit === 0n ? "0" : "1");
```

File: packages/runtime/src/interpreter/opcode.ts [hunk 3]
```diff
@@ -106,6 +106,28 @@ export class Op {
     return a;
   }
 
+  /**
+   * asserts if given index lies in 64 bit unsigned integer
+   * @param index Index
+   * @param line line number in TEAL file
+   */
+  assert64BitIndex (index: bigint, line: number): void {
+    if (index > 63n) {
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode.ts:115]: maybe we should add a constant `MAX_UINT6`?

File: packages/runtime/test/src/interpreter/opcode.ts [hunk 4]
```diff
@@ -44,3 +46,21 @@ describe("Teal Opcodes basic assertions", function () {
     );
   });
 });
+
+describe("parseToBigint", () => {
+  const op = new Op();
+
+  it("should parse to bigint", () => {
+    let res = op.parseToBigInt(["0"]);
+
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode.ts:55]:
```suggestion
```

File: packages/runtime/test/src/interpreter/opcode.ts [hunk 5]
```diff
@@ -44,3 +46,21 @@ describe("Teal Opcodes basic assertions", function () {
...
+    assert.equal(res, 0n);
+
+    res = op.parseToBigInt(["0", "1", "0"]);
+
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode.ts:59]:
```suggestion
```

File: packages/runtime/test/src/interpreter/opcode.ts [hunk 6]
```diff
@@ -44,3 +46,21 @@ describe("Teal Opcodes basic assertions", function () {
...
+    assert.equal(res, 2n);
+
+    res = op.parseToBigInt(["1", "0", "0", "0", "0"]);
+
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode.ts:63]:
```suggestion
```

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 7]
```diff
@@ -3211,4 +3212,250 @@ describe("Teal Opcodes", function () {
...
+      op.execute(stack);
+
+      assert.equal(stack.pop(), 1n);
+
+      stack.push(new Uint8Array([4, 0, 1])); // target
+      stack.push(6n); // index
+
+      op.execute(stack);
+
+      assert.equal(stack.pop(), 0n);
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:3415]:
```suggestion
      const op = new GetBit([], 0);
      stack.push(new Uint8Array([0, 128, 1])); // target
      stack.push(8n); // index
      op.execute(stack);
      assert.equal(stack.pop(), 1n);

      stack.push(new Uint8Array([1, 4, 1])); // target
      stack.push(23n); // index
      op.execute(stack);
      assert.equal(stack.pop(), 1n);

      stack.push(new Uint8Array([4, 0, 1])); // target
      stack.push(6n); // index
      op.execute(stack);
      assert.equal(stack.pop(), 0n);
```

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 8]
```diff
@@ -3211,4 +3212,250 @@ describe("Teal Opcodes", function () {
...
+    let stack: Stack<StackElem>;
+    this.beforeEach(() => { stack = new Stack<StackElem>(); });
+
+    it("should push correct bit to stack(uint64)", () => {
+      const op = new GetBit([], 0);
+      stack.push(8n); // target
+      stack.push(3n); // index
+
+      op.execute(stack);
+
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:3383]:
```suggestion
```

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 9]
```diff
@@ -3211,4 +3212,250 @@ describe("Teal Opcodes", function () {
...
+      assert.equal(stack.pop(), 1n);
+
+      stack.push(8n); // target
+      stack.push(0n); // index
+
+      op.execute(stack);
+
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:3390]:
```suggestion
```

File: packages/runtime/src/interpreter/opcode.ts [hunk 10]
```diff
@@ -176,4 +198,19 @@ export class Op {
     }
     return stack;
   }
+
+  /**
+   * Parses binary string into bigint
+   * @param binary Binary string array
+   */
+  parseToBigInt (binary: string[]): bigint {
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode.ts:206]: Can we also update this function to take a string as well? Eg. `parseToBigInt('101') => 5n`

Comment (ratik21) [reply packages/runtime/src/interpreter/opcode.ts:206]: Also `parseToBigInt` sounds a bit ambigous. Maybe we could rename to `parseBinaryStrToBigInt` ?

File: packages/runtime/src/interpreter/opcode.ts [hunk 11]
```diff
@@ -176,4 +198,19 @@ export class Op {
...
+    let res = 0n;
+    for (let i = 0; i < binary.length; ++i) {
+      if (binary[i] === '1') {
+        const val = binary.length - 1 - i;
+        res += 2n ** BigInt(val);
+      }
+    }
+    return res;
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode.ts:214]: How about moving this function to `lib/parsing.ts`?

File: packages/runtime/test/src/interpreter/opcode.ts [hunk 12]
```diff
@@ -44,3 +46,21 @@ describe("Teal Opcodes basic assertions", function () {
...
+    assert.equal(res, 16n);
+  });
+});
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode.ts:66]: please also add test for `op.parseToBigInt(["1"]);`, `op.parseToBigInt(["0", "0"])`

File: packages/runtime/src/interpreter/opcode-list.ts (hunk 1 above)
Comment (amityadav0) [reply packages/runtime/src/interpreter/opcode-list.ts:2269]: ok

File: packages/runtime/src/interpreter/opcode.ts (hunk 3 above)
Comment (amityadav0) [reply packages/runtime/src/interpreter/opcode.ts:115]: yes

File: packages/runtime/src/interpreter/opcode.ts (hunk 10 above)
Comment (amityadav0) [reply packages/runtime/src/interpreter/opcode.ts:206]: yes

File: packages/runtime/test/src/interpreter/opcode.ts (hunk 12 above)
Comment (amityadav0) [reply packages/runtime/test/src/interpreter/opcode.ts:63]: ok

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 13]
```diff
@@ -2259,3 +2259,104 @@ export class Swap extends Op {
...
 }
+
+/**
+ * bit indexing begins with low-order bits in integers.
+ * Setting bit 4 to 1 on the integer 0 yields 16 (int 0x0010, or 2^4).
+ * Indexing begins in the first bytes of a byte-string
+ * (as seen in getbyte and substring). Setting bits 0 through 11 to 1
+ * in a 4 byte-array of 0s yields byte 0xfff00000
+ * Pops: ... stack, {any A}, {uint64 B}, {uint64 C}
+ * Pushes: uint64
```
Comment (ratik21) [inline packages/runtime/src/interpreter/opcode-list.ts:2270]:
```suggestion
 * Pops from stack: [ ... stack, {any A}, {uint64 B}, {uint64 C} ] 
 * Pushes to stack: [ ...stack, uint64 ]
```

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 14]
```diff
@@ -3223,4 +3224,242 @@ describe("Teal Opcodes", function () {
...
+      assert.equal(stack.pop(), 14n);
+
+      stack.push(0n);
+      stack.push(63n);
+      stack.push(1n);
+
+      op.execute(stack);
+
+      assert.equal(stack.length(), 1);
+      assert.equal(stack.pop(), 2n ** 63n);
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:3268]: Can we also add a test where target is `MAX_UINT64`?

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 15]
```diff
@@ -3223,4 +3224,242 @@ describe("Teal Opcodes", function () {
...
+      stack.push(4n);
+      stack.push(20n);
+
+      expectRuntimeError(
+        () => op.execute(stack),
+        RUNTIME_ERRORS.TEAL.SET_BIT_VALUE_ERROR
+      );
+    });
+
+    it("should panic if set bit index is greater than 63", () => {
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:3318]:
```suggestion
    it("should panic if set bit index is greater than 63 and target is uint64", () => {
```

File: packages/runtime/test/src/interpreter/opcode-list.ts [hunk 16]
```diff
@@ -3223,4 +3224,242 @@ describe("Teal Opcodes", function () {
...
+    it("should panic if index bit in out of bytes array", () => {
+      const op = new SetBit([], 0);
+      stack.push(new Uint8Array([0, 0, 0])); // target
+      stack.push(80n); // index
+      stack.push(1n); // bit
+
+      expectRuntimeError(
+        () => op.execute(stack),
+        RUNTIME_ERRORS.TEAL.SET_BIT_INDEX_BYTES_ERROR
+      );
```
Comment (ratik21) [inline packages/runtime/test/src/interpreter/opcode-list.ts:3381]: please add a corner case here. target = `new Uint8Array(8).fill(0)`; `index = 64 * 8 + 1`

File: packages/runtime/src/errors/errors-list.ts [hunk 17]
```diff
@@ -219,6 +219,24 @@ by an index that does not exist.`
...
+  },
+  SET_BIT_VALUE_ERROR: {
+    number: 1032,
+    message: "set bit value is greater than 1. [error-line: %line%]",
+    title: 'Bit value error',
+    description: `Bit value error`
+  },
+  SET_BIT_INDEX_ERROR: {
+    number: 1033,
+    message: "set bit index is greater than 63 with Uint. [error-line: %line%]",
```
Comment (ratik21) [inline packages/runtime/src/errors/errors-list.ts:231]: can we display index fetched from stack (> 63) in the error-msg as well?

File: packages/runtime/src/errors/errors-list.ts [hunk 18]
```diff
@@ -219,6 +219,24 @@ by an index that does not exist.`
...
+    title: 'set bit index error',
+    description: `set bit index error`
+  },
+  SET_BIT_INDEX_BYTES_ERROR: {
+    number: 1034,
+    message: "set bit index is beyond bytes length. [error-line: %line%]",
```
Comment (ratik21) [inline packages/runtime/src/errors/errors-list.ts:237]: Same as above, i think we should display index in the error message.

File: packages/runtime/src/lib/parsing.ts [hunk 19]
```diff
@@ -244,3 +244,18 @@ export function getEncoding (args: string[], line: number): [string, EncodingTyp
     throw new RuntimeError(RUNTIME_ERRORS.TEAL.UNKOWN_DECODE_TYPE, { val: args[0], line: line });
   }
 }
+
+/**
+ * Parses binary string into bigint
```
Comment (ratik21) [inline packages/runtime/src/lib/parsing.ts:249]:
```suggestion
 * Parses binary string into bigint. Eg '101' OR ['1', '0', '1'] => 5n
```

File: packages/runtime/src/interpreter/opcode-list.ts [hunk 20]
```diff
@@ -2259,3 +2259,104 @@ export class Swap extends Op {
...
+    const target = stack.pop();
+
+    if (typeof target === "bigint") {
+      this.assert64BitIndex(index, this.line);
+      const binaryStr = target.toString(2);
+      const size = binaryStr.length;
+      stack.push(BigInt(binaryStr[size - Number(index) - 1]));
+    } else {
+      const byteIndex = Math.floor(Number(index) / 8);
+      this.assertBytesIndex(byteIndex, target.length, this.line);
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/runtime/src/interpreter/opcode-list.ts:2354]: But it is more optimised, instead of sending a while array sending a value is faster, but if it's about `intuition` I can do it.

**Review by ratik21 — APPROVED** [review]
